File: packages/core/src/schema.ts
import { IntentBlock, IntentDocument, BlockType } from "./types";
import { flattenBlocks } from "./utils";

export interface PropertySchema {
  type: "string" | "number" | "boolean" | "date" | "enum" | "url" | "email";
  required?: boolean;
  default?: string | number | boolean;
  enumValues?: string[]; // for enum type
  pattern?: string; // regex for string validation
  min?: number; // for numbers
  max?: number; // for numbers
  format?: "iso-date" | "iso-datetime" | "time" | "url" | "email";
}

export interface BlockSchema {
  type: BlockType;
  content?: {
    required?: boolean;
    minLength?: number;
    maxLength?: number;
    pattern?: string;
  };
  properties?: Record<string, PropertySchema>;
  allowUnknownProperties?: boolean;
}

export interface DocumentSchema {
  name: string;
  description?: string;
  requiredBlocks?: BlockType[];
  blockSchemas?: Record<string, BlockSchema>; // key is block type
  allowUnknownBlocks?: boolean;
}

export interface ValidationError {
  blockId: string;
  blockType: string;
  field: string;
  message: string;
  severity: "error" | "warning";
}

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
  warnings: ValidationError[];
}

// Predefined schemas for common use cases
export const PREDEFINED_SCHEMAS: Record<string, DocumentSchema> = {
  project: {
    name: "project",
    description: "Project planning document",
    requiredBlocks: ["title"],
    blockSchemas: {
      task: {
        type: "task",
        properties: {
          owner: { type: "string" },
          due: { type: "date", format: "iso-date" },
          priority: { type: "enum", enumValues: ["low", "medium", "high"] },
        },
        allowUnknownProperties: true,
      },
      done: {
        type: "done",
        properties: {
          time: { type: "string", format: "time" },
        },
      },
    },
    allowUnknownBlocks: true,
  },

  meeting: {
    name: "meeting",
    description: "Meeting notes document",
    requiredBlocks: ["title", "section"],
    blockSchemas: {
      note: {
        type: "note",
        content: { required: true },
      },
      ask: {
        type: "ask",
        content: { required: true },
      },
      task: {
        type: "task",
        properties: {
          owner: { type: "string", required: true },
          due: { type: "date", format: "iso-date", required: true },
        },
      },
    },
    allowUnknownBlocks: true,
  },

  article: {
    name: "article",
    description: "Blog post or article",
    requiredBlocks: ["title", "summary"],
    blockSchemas: {
      image: {
        type: "image",
        properties: {
          at: { type: "url", required: true },
          caption: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      link: {
        type: "link",
        content: { required: true, minLength: 1 },
        properties: {
          to: { type: "url", required: true },
        },
      },
      section: {
        type: "section",
        content: { required: true, minLength: 1 },
      },
    },
    allowUnknownBlocks: false,
  },

  checklist: {
    name: "checklist",
    description: "Simple task checklist",
    requiredBlocks: ["title"],
    blockSchemas: {
      task: {
        type: "task",
        content: { required: true },
      },
      done: {
        type: "done",
        content: { required: true },
      },
    },
    allowUnknownBlocks: false,
  },

  agentic: {
    name: "agentic",
    description: "Agentic workflow document (v2.0+)",
    requiredBlocks: ["title"],
    blockSchemas: {
      step: {
        type: "step",
        content: { required: true },
        properties: {
          tool: { type: "string" },
          status: {
            type: "enum",
            enumValues: [
              "pending",
              "running",
              "blocked",
              "failed",
              "skipped",
              "cancelled",
              "done",
            ],
          },
          id: { type: "string" },
          depends: { type: "string" },
          timeout: { type: "number" },
          priority: { type: "number" },
          retries: { type: "number" },
        },
        allowUnknownProperties: true,
      },
      decision: {
        type: "decision",
        content: { required: true },
        properties: {
          if: { type: "string" },
          then: { type: "string" },
          else: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      trigger: {
        type: "trigger",
        content: { required: true },
        properties: {
          event: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      loop: {
        type: "loop",
        content: { required: true },
        properties: {
          over: { type: "string" },
          do: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      retry: {
        type: "retry",
        content: { required: true },
        properties: {
          max: { type: "number" },
          delay: { type: "number" },
          backoff: {
            type: "enum",
            enumValues: ["linear", "exponential", "fixed"],
          },
        },
        allowUnknownProperties: true,
      },
      wait: {
        type: "wait",
        content: { required: true },
        properties: {
          timeout: { type: "string" },
          fallback: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      parallel: {
        type: "parallel",
        content: { required: true },
        properties: {
          steps: { type: "string" },
          timeout: { type: "number" },
        },
        allowUnknownProperties: true,
      },
      handoff: {
        type: "handoff",
        content: { required: true },
        properties: {
          from: { type: "string" },
          to: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      result: {
        type: "result",
        content: { required: true },
        properties: {
          status: {
            type: "enum",
            enumValues: ["success", "error", "failure"],
          },
          code: { type: "string" },
          data: { type: "string" },
        },
        allowUnknownProperties: true,
      },
      status: {
        type: "status",
        content: { required: true },
        properties: {
          phase: { type: "string" },
          level: { type: "string" },
          updated: { type: "string" },
        },
        allowUnknownProperties: true,
      },
    },
    allowUnknownBlocks: true,
  },
};

function validateProperty(
  value: unknown,
  schema: PropertySchema,
  fieldName: string,
): ValidationError[] {
  const errors: ValidationError[] = [];
  const strValue = String(value);

  // Required check
  if (
    schema.required &&
    (value === undefined || value === null || value === "")
  ) {
    errors.push({
      blockId: "",
      blockType: "",
      field: fieldName,
      message: `Property "${fieldName}" is required`,
      severity: "error",
    });
    return errors;
  }

  // Skip further validation if value is empty and not required
  if (value === undefined || value === null || value === "") {
    return errors;
  }

  switch (schema.type) {
    case "number":
      if (isNaN(Number(value))) {
        errors.push({
          blockId: "",
          blockType: "",
          field: fieldName,
          message: `Property "${fieldName}" must be a number`,
          severity: "error",
        });
      } else {
        const num = Number(value);
        if (schema.min !== undefined && num < schema.min) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be >= ${schema.min}`,
            severity: "error",
          });
        }
        if (schema.max !== undefined && num > schema.max) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be <= ${schema.max}`,
            severity: "error",
          });
        }
      }
      break;

    case "date":
      const date = new Date(strValue);
      if (isNaN(date.getTime())) {
        // Try parsing as ISO date specifically
        const isoDateRegex = /^\d{4}-\d{2}-\d{2}$/;
        if (schema.format === "iso-date" && !isoDateRegex.test(strValue)) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be in ISO date format (YYYY-MM-DD)`,
            severity: "error",
          });
        } else if (!isoDateRegex.test(strValue) && !strValue.includes("T")) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be a valid date`,
            severity: "error",
          });
        }
      }
      break;

    case "enum":
      if (schema.enumValues && !schema.enumValues.includes(strValue)) {
        errors.push({
          blockId: "",
          blockType: "",
          field: fieldName,
          message: `Property "${fieldName}" must be one of: ${schema.enumValues.join(", ")}`,
          severity: "error",
        });
      }
      break;

    case "url":
      try {
        new URL(strValue);
      } catch {
        // Allow relative URLs
        if (!strValue.startsWith("/") && !strValue.startsWith("./")) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be a valid URL`,
            severity: "error",
          });
        }
      }
      break;

    case "email":
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
      if (!emailRegex.test(strValue)) {
        errors.push({
          blockId: "",
          blockType: "",
          field: fieldName,
          message: `Property "${fieldName}" must be a valid email address`,
          severity: "error",
        });
      }
      break;

    case "string":
      if (schema.pattern) {
        const regex = new RegExp(schema.pattern);
        if (!regex.test(strValue)) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" does not match required pattern`,
            severity: "error",
          });
        }
      }
      break;
  }

  // Format validation
  if (schema.format) {
    switch (schema.format) {
      case "iso-date":
        if (!/^\d{4}-\d{2}-\d{2}$/.test(strValue)) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be ISO date (YYYY-MM-DD)`,
            severity: "error",
          });
        }
        break;
      case "time":
        if (!/^\d{1,2}:\d{2}(?::\d{2})?\s*(?:AM|PM)?$/i.test(strValue)) {
          errors.push({
            blockId: "",
            blockType: "",
            field: fieldName,
            message: `Property "${fieldName}" must be a valid time`,
            severity: "warning",
          });
        }
        break;
    }
  }

  return errors;
}

function validateBlock(
  block: IntentBlock,
  schema: BlockSchema,
  documentSchema: DocumentSchema,
): ValidationError[] {
  const errors: ValidationError[] = [];

  // Validate content if schema defined
  if (schema.content) {
    if (schema.content.required && (!block.content || block.content === "")) {
      errors.push({
        blockId: block.id,
        blockType: block.type,
        field: "content",
        message: `Block "${block.type}" requires content`,
        severity: "error",
      });
    }
    if (block.content) {
      if (
        schema.content.minLength !== undefined &&
        block.content.length < schema.content.minLength
      ) {
        errors.push({
          blockId: block.id,
          blockType: block.type,
          field: "content",
          message: `Content must be at least ${schema.content.minLength} characters`,
          severity: "error",
        });
      }
      if (
        schema.content.maxLength !== undefined &&
        block.content.length > schema.content.maxLength
      ) {
        errors.push({
          blockId: block.id,
          blockType: block.type,
          field: "content",
          message: `Content must be at most ${schema.content.maxLength} characters`,
          severity: "error",
        });
      }
      if (schema.content.pattern) {
        const regex = new RegExp(schema.content.pattern);
        if (!regex.test(block.content)) {
          errors.push({
            blockId: block.id,
            blockType: block.type,
            field: "content",
            message: `Content does not match required pattern`,
            severity: "error",
          });
        }
      }
    }
  }

  // Validate properties
  if (schema.properties) {
    for (const [propName, propSchema] of Object.entries(schema.properties)) {
      const value = block.properties?.[propName];
      const propErrors = validateProperty(value, propSchema, propName);
      errors.push(
        ...propErrors.map((e) => ({
          ...e,
          blockId: block.id,
          blockType: block.type,
        })),
      );
    }

    // Check for unknown properties
    if (schema.allowUnknownProperties === false && block.properties) {
      const knownProps = Object.keys(schema.properties);
      for (const propName of Object.keys(block.properties)) {
        if (!knownProps.includes(propName)) {
          errors.push({
            blockId: block.id,
            blockType: block.type,
            field: propName,
            message: `Unknown property "${propName}" for block type "${block.type}"`,
            severity: "warning",
          });
        }
      }
    }
  }

  return errors;
}

/**
 * Validate an IntentDocument against a schema
 */
export function validateDocument(
  document: IntentDocument,
  schema: DocumentSchema | string,
): ValidationResult {
  // If string provided, look up predefined schema
  const docSchema =
    typeof schema === "string" ? PREDEFINED_SCHEMAS[schema] : schema;

  if (!docSchema) {
    return {
      valid: false,
      errors: [
        {
          blockId: "",
          blockType: "",
          field: "schema",
          message: `Unknown schema "${schema}"`,
          severity: "error",
        },
      ],
      warnings: [],
    };
  }

  const allBlocks = flattenBlocks(document.blocks);
  const errors: ValidationError[] = [];
  const warnings: ValidationError[] = [];

  // Check required blocks
  if (docSchema.requiredBlocks) {
    const presentTypes = new Set(allBlocks.map((b) => b.type));
    for (const required of docSchema.requiredBlocks) {
      if (!presentTypes.has(required)) {
        errors.push({
          blockId: "",
          blockType: "",
          field: "blocks",
          message: `Document requires at least one "${required}" block`,
          severity: "error",
        });
      }
    }
  }

  // Check unknown blocks
  if (docSchema.allowUnknownBlocks === false) {
    const allowedTypes = new Set(Object.keys(docSchema.blockSchemas || {}));
    // Always allow structural blocks
    allowedTypes.add("title");
    allowedTypes.add("summary");
    allowedTypes.add("section");
    allowedTypes.add("sub");
    allowedTypes.add("divider");

    for (const block of allBlocks) {
      if (!allowedTypes.has(block.type)) {
        errors.push({
          blockId: block.id,
          blockType: block.type,
          field: "type",
          message: `Block type "${block.type}" is not allowed in "${docSchema.name}" documents`,
          severity: "error",
        });
      }
    }
  }

  // Validate blocks against their schemas
  if (docSchema.blockSchemas) {
    for (const block of allBlocks) {
      const blockSchema = docSchema.blockSchemas[block.type];
      if (blockSchema) {
        const blockErrors = validateBlock(block, blockSchema, docSchema);
        for (const e of blockErrors) {
          if (e.severity === "error") errors.push(e);
          else warnings.push(e);
        }
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}

/**
 * Create a custom schema
 */
export function createSchema(
  config: Omit<DocumentSchema, "name"> & { name: string },
): DocumentSchema {
  return {
    allowUnknownBlocks: true,
    ...config,
  };
}

/**
 * Format validation results for CLI output
 */
export function formatValidationResult(result: ValidationResult): string {
  if (result.valid && result.warnings.length === 0) {
    return "✓ Document is valid";
  }

  const lines: string[] = [];

  if (!result.valid) {
    lines.push(`✗ Validation failed with ${result.errors.length} error(s):`);
    for (const e of result.errors) {
      lines.push(
        `  [${e.severity.toUpperCase()}] ${e.blockType || "document"}: ${e.message}`,
      );
    }
  }

  if (result.warnings.length > 0) {
    lines.push(`⚠ ${result.warnings.length} warning(s):`);
    for (const w of result.warnings) {
      lines.push(`  [WARN] ${w.blockType}: ${w.message}`);
    }
  }

  if (result.valid && result.warnings.length === 0) {
    lines.push("✓ Document is valid");
  }

  return lines.join("\n");
}

File: packages/rust/src/schema.rs
//! Document schema validation parity surface.

use std::collections::HashMap;

use once_cell::sync::Lazy;

use crate::types::{IntentBlock, IntentDocument};

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum PropertyType {
    #[default]
    String,
    Number,
    Boolean,
    Date,
    Enum,
    Url,
    Email,
}

#[derive(Debug, Clone, Default)]
pub struct PropertySchema {
    pub property_type: PropertyType,
    pub required: bool,
    pub enum_values: Vec<String>,
    pub min: Option<f64>,
    pub max: Option<f64>,
    pub format: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct ContentSchema {
    pub required: bool,
    pub min_length: Option<usize>,
    pub max_length: Option<usize>,
}

#[derive(Debug, Clone, Default)]
pub struct BlockSchema {
    pub block_type: String,
    pub content: Option<ContentSchema>,
    pub properties: HashMap<String, PropertySchema>,
    pub allow_unknown_properties: bool,
}

#[derive(Debug, Clone, Default)]
pub struct DocumentSchema {
    pub name: String,
    pub description: Option<String>,
    pub required_blocks: Vec<String>,
    pub block_schemas: HashMap<String, BlockSchema>,
    pub allow_unknown_blocks: bool,
}

#[derive(Debug, Clone, Default)]
pub struct ValidationError {
    pub block_id: String,
    pub block_type: String,
    pub field: String,
    pub message: String,
    pub severity: String,
}

#[derive(Debug, Clone, Default)]
pub struct ValidationResult {
    pub valid: bool,
    pub errors: Vec<ValidationError>,
    pub warnings: Vec<ValidationError>,
}

pub static PREDEFINED_SCHEMAS: Lazy<HashMap<String, DocumentSchema>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("project".to_string(), project_schema());
    m.insert("meeting".to_string(), meeting_schema());
    m.insert("article".to_string(), article_schema());
    m.insert("checklist".to_string(), checklist_schema());
    m.insert("agentic".to_string(), agentic_schema());
    m
});

pub fn create_schema(name: &str) -> Option<DocumentSchema> {
    PREDEFINED_SCHEMAS.get(name).cloned()
}

pub fn validate_document(document: &IntentDocument, schema: &DocumentSchema) -> ValidationResult {
    let mut out = ValidationResult::default();
    let blocks = &document.blocks;

    for req in &schema.required_blocks {
        if !blocks.iter().any(|b| b.block_type == *req) {
            out.errors.push(ValidationError {
                block_id: String::new(),
                block_type: req.clone(),
                field: "type".to_string(),
                message: format!("Required block type missing: {req}"),
                severity: "error".to_string(),
            });
        }
    }

    for b in blocks {
        if let Some(bs) = schema.block_schemas.get(&b.block_type) {
            validate_block(b, bs, &mut out);
        } else if !schema.allow_unknown_blocks {
            out.warnings.push(ValidationError {
                block_id: b.id.clone(),
                block_type: b.block_type.clone(),
                field: "type".to_string(),
                message: format!("Unknown block type for schema: {}", b.block_type),
                severity: "warning".to_string(),
            });
        }
    }

    out.valid = out.errors.is_empty();
    out
}

pub fn format_validation_result(result: &ValidationResult) -> String {
    if result.valid && result.warnings.is_empty() {
        return "Validation passed".to_string();
    }
    let mut lines = Vec::new();
    if !result.errors.is_empty() {
        lines.push(format!("Errors: {}", result.errors.len()));
        for e in &result.errors {
            lines.push(format!("- {} [{}:{}]", e.message, e.block_type, e.field));
        }
    }
    if !result.warnings.is_empty() {
        lines.push(format!("Warnings: {}", result.warnings.len()));
        for w in &result.warnings {
            lines.push(format!("- {} [{}:{}]", w.message, w.block_type, w.field));
        }
    }
    lines.join("\n")
}

fn validate_block(block: &IntentBlock, schema: &BlockSchema, out: &mut ValidationResult) {
    if let Some(c) = &schema.content {
        let len = block.content.chars().count();
        if c.required && block.content.trim().is_empty() {
            out.errors.push(ValidationError {
                block_id: block.id.clone(),
                block_type: block.block_type.clone(),
                field: "content".to_string(),
                message: "Content is required".to_string(),
                severity: "error".to_string(),
            });
        }
        if let Some(min) = c.min_length {
            if len < min {
                out.errors.push(ValidationError {
                    block_id: block.id.clone(),
                    block_type: block.block_type.clone(),
                    field: "content".to_string(),
                    message: format!("Content must be at least {min} characters"),
                    severity: "error".to_string(),
                });
            }
        }
        if let Some(max) = c.max_length {
            if len > max {
                out.errors.push(ValidationError {
                    block_id: block.id.clone(),
                    block_type: block.block_type.clone(),
                    field: "content".to_string(),
                    message: format!("Content must be at most {max} characters"),
                    severity: "error".to_string(),
                });
            }
        }
    }

    let props = block.properties.as_ref().cloned().unwrap_or_default();
    for (key, ps) in &schema.properties {
        let v = props.get(key);
        if ps.required && v.is_none() {
            out.errors.push(ValidationError {
                block_id: block.id.clone(),
                block_type: block.block_type.clone(),
                field: key.clone(),
                message: format!("Property \"{key}\" is required"),
                severity: "error".to_string(),
            });
            continue;
        }
        if let Some(v) = v {
            validate_property(block, key, v, ps, out);
        }
    }

    if !schema.allow_unknown_properties {
        for key in props.keys() {
            if !schema.properties.contains_key(key) {
                out.warnings.push(ValidationError {
                    block_id: block.id.clone(),
                    block_type: block.block_type.clone(),
                    field: key.clone(),
                    message: format!("Unknown property for schema: {key}"),
                    severity: "warning".to_string(),
                });
            }
        }
    }
}

fn validate_property(
    block: &IntentBlock,
    field: &str,
    value: &str,
    schema: &PropertySchema,
    out: &mut ValidationResult,
) {
    match schema.property_type {
        PropertyType::Number => {
            if let Ok(n) = value.parse::<f64>() {
                if let Some(min) = schema.min {
                    if n < min {
                        out.errors
                            .push(prop_err(block, field, format!("must be >= {min}")));
                    }
                }
                if let Some(max) = schema.max {
                    if n > max {
                        out.errors
                            .push(prop_err(block, field, format!("must be <= {max}")));
                    }
                }
            } else {
                out.errors
                    .push(prop_err(block, field, "must be a number".to_string()));
            }
        }
        PropertyType::Enum => {
            if !schema.enum_values.is_empty() && !schema.enum_values.iter().any(|v| v == value) {
                out.errors.push(prop_err(
                    block,
                    field,
                    format!("must be one of: {}", schema.enum_values.join(", ")),
                ));
            }
        }
        PropertyType::Url => {
            let low = value.to_ascii_lowercase();
            if !(low.starts_with("http://")
                || low.starts_with("https://")
                || low.starts_with("/")
                || low.starts_with("./")
                || low.starts_with("../"))
            {
                out.errors
                    .push(prop_err(block, field, "must be a URL".to_string()));
            }
        }
        PropertyType::Email => {
            if !value.contains('@') {
                out.errors
                    .push(prop_err(block, field, "must be an email".to_string()));
            }
        }
        PropertyType::Date => {
            if let Some(fmt) = &schema.format {
                if fmt == "iso-date" {
                    let ok = value.len() == 10
                        && value.chars().nth(4) == Some('-')
                        && value.chars().nth(7) == Some('-');
                    if !ok {
                        out.errors.push(prop_err(
                            block,
                            field,
                            "must be in ISO date format (YYYY-MM-DD)".to_string(),
                        ));
                    }
                }
            }
        }
        PropertyType::String | PropertyType::Boolean => {}
    }
}

fn prop_err(block: &IntentBlock, field: &str, message: String) -> ValidationError {
    ValidationError {
        block_id: block.id.clone(),
        block_type: block.block_type.clone(),
        field: field.to_string(),
        message: format!("Property \"{field}\" {message}"),
        severity: "error".to_string(),
    }
}

fn project_schema() -> DocumentSchema {
    let mut task_props = HashMap::new();
    task_props.insert(
        "priority".to_string(),
        PropertySchema {
            property_type: PropertyType::Enum,
            enum_values: vec!["low".to_string(), "medium".to_string(), "high".to_string()],
            ..Default::default()
        },
    );
    task_props.insert(
        "due".to_string(),
        PropertySchema {
            property_type: PropertyType::Date,
            format: Some("iso-date".to_string()),
            ..Default::default()
        },
    );

    let mut block_schemas = HashMap::new();
    block_schemas.insert(
        "task".to_string(),
        BlockSchema {
            block_type: "task".to_string(),
            properties: task_props,
            allow_unknown_properties: true,
            ..Default::default()
        },
    );

    DocumentSchema {
        name: "project".to_string(),
        description: Some("Project planning document".to_string()),
        required_blocks: vec!["title".to_string()],
        block_schemas,
        allow_unknown_blocks: true,
    }
}

fn meeting_schema() -> DocumentSchema {
    DocumentSchema {
        name: "meeting".to_string(),
        description: Some("Meeting notes document".to_string()),
        required_blocks: vec!["title".to_string(), "section".to_string()],
        allow_unknown_blocks: true,
        ..Default::default()
    }
}

fn article_schema() -> DocumentSchema {
    DocumentSchema {
        name: "article".to_string(),
        description: Some("Article document".to_string()),
        required_blocks: vec!["title".to_string(), "summary".to_string()],
        allow_unknown_blocks: false,
        ..Default::default()
    }
}

fn checklist_schema() -> DocumentSchema {
    DocumentSchema {
        name: "checklist".to_string(),
        description: Some("Checklist document".to_string()),
        required_blocks: vec!["title".to_string()],
        allow_unknown_blocks: false,
        ..Default::default()
    }
}

fn agentic_schema() -> DocumentSchema {
    DocumentSchema {
        name: "agentic".to_string(),
        description: Some("Agentic workflow document".to_string()),
        required_blocks: vec!["title".to_string()],
        allow_unknown_blocks: true,
        ..Default::default()
    }
}
